kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -0,0 +1,331 @@
+    private final MappingIterator<T> input;
+
+    private InputStreamingResponse(MappingIterator<T> input) {
+      this.input = requireNonNull(input);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return input.hasNext();
+    }
+
+    @Override
+    public CompletableFuture<T> next() {
+      try {
+        return CompletableFuture.completedFuture(input.nextValue());
+      } catch (Throwable e) {
+        return CompletableFutures.failedFuture(e);
+      }
+    }
+  }
+
+  private static final class ComposingStreamingResponse<I, O> extends StreamingResponse<O> {
+    private final StreamingResponse<I> input;
+    private final Function<? super I, ? extends CompletableFuture<O>> transform;
+
+    private ComposingStreamingResponse(
+        StreamingResponse<I> input, Function<? super I, ? extends CompletableFuture<O>> transform) {
+      this.input = requireNonNull(input);
+      this.transform = requireNonNull(transform);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return input.hasNext();
+    }
+
+    @Override
+    public CompletableFuture<O> next() {
+      return input.next().thenCompose(transform);
+    }
+  }
+
+  private static final class AsyncResponseQueue {
+    private static final String CHUNK_SEPARATOR = "\r\n";
+
+    private final ChunkedOutput<ResultOrError> sink;
+
+    // tail is the end of a linked list of completable futures. The futures are tied together by
+    // allOf completion handles. For example, let's say we push 3 futures into the queue:
+    //
+    // 1. tail = (null) // empty queue
+    // 2. tail = then(allOf((null), f_1), write(f_1)) // -> f_1
+    // 3. tail = then(allOf(then(allOf((null), f_1), write(f_1)), f_2), write(f_2)) // -> f_1 -> f_2
+    //
+    // The chaining of the futures guarantees that the write(f) callbacks are called in the order
+    // that the futures got pushed into the queue. Also, due to the way allOf is implemented, it
+    // makes sure that, as soon as f_0..i is completed, the whole
+    // then(allOf((f_i-1), f_i), write(f_i)) monad is made available for garbage collection.
+    private CompletableFuture<Void> tail;
+
+    private AsyncResponseQueue() {
+      sink = new ChunkedOutput<>(ResultOrError.class, CHUNK_SEPARATOR);
+      tail = CompletableFuture.completedFuture(null);
+    }
+
+    private void asyncResume(AsyncResponse asyncResponse) {
+      asyncResponse.resume(Response.ok(sink).build());
+    }
+
+    private void push(CompletableFuture<ResultOrError> result) {
+      tail =
+          CompletableFuture.allOf(tail, result)
+              .thenApply(
+                  unused -> {
+                    try {
+                      sink.write(result.join());
+                    } catch (IOException e) {
+                      log.error("Error when writing streaming result to response channel.", e);
+                    }
+                    return null;
+                  });
+    }
```

Comment:
Have you tested this manually, and can you please write a unit test for it?

Correct me if I'm wrong, but I'm concerned about the following:
1. A single request from the middle of a multi-request chunk getting an error during the produce handling on the broker side and therefore having its future (`result`) complete exceptionally.
2. The `CompletableFuture#allOf` tail wrapper corresponding to that `result` completing exceptionally too, and with that making any of the subsequent `CompletableFuture#allOf` tail wrappers to complete exceptionally too, even if the other half of what they wrap (their `result`s, the actual produce futures here) completes normally. 
3. The `thenApply` calls for these tail wrappers not being called (as `thenApply` is called for normal completions), despite their `result` containing a valid result that could be written back to the client.

In addition I don't think we're logging anything for the exceptions causing exceptional completion.

Maybe we should use `whenComplete` here, and in its body, specifically catching and logging for `CompletionException`s from `result.join()`?

Reply:
I just noticed that the `handleNext` result is what is passed down as the `result` future here, so the above should not be a concern.

Reply:
I have not written a test for StreamingResponse because it ultimately terminates into a ChunkedOutput sent to AsyncResponse. Although it's possible to mock the AsyncResponse, we can't get anything out of the ChunkedOutput (it's not a very good interface). I will try to improve the testability of this class later, but I prefer to delay it to a later PR.

Reply:
Agreed.